### PR TITLE
Add low level control with pybind11

### PR DESCRIPTION
**Pull request by DezhongT**
This branch adds the low level control for positions, twists, and curvatures. Therein, we can use the python codes to create the controller for generating control signals in the action space and then step the simulation in C++. We might also need to think about how to define the observations and get the observations.

### CODE REVIEW

File: py_examples/helix_case/helixExample.py
Comment (QuantuMope) [inline py_examples/helix_case/helixExample.py:1]: This example file should not be modified. This is a file representing the experiment done in the original DisMech manuscript.

File: py_examples/helix_case/helixExample.py [hunk 1]
```diff
@@ -51,7 +54,14 @@
...
 # Initialize and run the simulation
 sim_manager.initialize(sys.argv)
+time = 0.0
 while not sim_manager.simulation_completed():
-    sim_manager.step_simulation()
+    # sim_manager.step_simulation()
+
+    input_dict = {"position" : np.array([[0, 0, time * 0.01, 0.0, 0.0], [0, 1, time * 0.01, 0.0, 0.0]])}
+    sim_manager.step_simulation(input_dict)
+    time += sim_params.dt
```
Comment (QuantuMope) [inline py_examples/helix_case/helixExample.py:67]: Running
```
pip install pre-commit
pre-commit install
pre-commit run --all-files
```
will fix indentations eof issues like this.

File: src/rod_mechanics/softRobots.cpp [hunk 2]
```diff
@@ -57,6 +57,71 @@ void softRobots::applyInitialVelocities(int limb_idx, const vector<Vector3d> &ve
...
+
+        // get material frames   
+        // Vector3d curvature = Curvatures.row(i).segment(2, 3).transpose();
+        double cx = Curvatures(i, 2);
+        double cy = Curvatures(i, 3);
+
+        // Vector3d m1e = limbs[limb_idx]->m1.row(node_idx - 1);
+        // Vector3d m2e = limbs[limb_idx]->m2.row(node_idx - 1);
+        // Vector3d m1f = limbs[limb_idx]->m1.row(node_idx);
+        // Vector3d m2f = limbs[limb_idx]->m2.row(node_idx);
```
Comment (QuantuMope) [inline src/rod_mechanics/softRobots.cpp:119]: Remove commented out code.

File: src/rod_mechanics/softRobots.cpp [hunk 3]
```diff
@@ -57,6 +57,71 @@ void softRobots::applyInitialVelocities(int limb_idx, const vector<Vector3d> &ve
...
+    for (int i = 0; i < Curvatures.rows(); i++) {
+        // the first col is limb_idx, second col is node_idx, third col is x, fourth col is y, fifth col is z 
+        int limb_idx = Curvatures(i, 0);
+        int node_idx = Curvatures(i, 1);
+        if (limb_idx >= limbs.size() || node_idx >= limbs[limb_idx]->ne || node_idx < 1) {
+            throw runtime_error("Invalid limb_idx or edge_idx given!");
+        }
+
+        // get material frames   
+        // Vector3d curvature = Curvatures.row(i).segment(2, 3).transpose();
```
Comment (QuantuMope) [inline src/rod_mechanics/softRobots.cpp:112]: Remove commented out code.

File: src/rod_mechanics/softRobots.cpp [hunk 4]
```diff
@@ -57,6 +57,71 @@ void softRobots::applyInitialVelocities(int limb_idx, const vector<Vector3d> &ve
...
+        if (limb_idx >= limbs.size() || edge_idx >= limbs[limb_idx]->ne) {
+            throw runtime_error("Invalid limb_idx or edge_idx given!");
+        }
+        
+        double theta = twists(i, 2);
+        limbs[limb_idx]->x(4*edge_idx + 3) += theta;
+    }
+}
+
+void softRobots::applyCurvatureBC(const MatrixXd &Curvatures) {
```
Comment (QuantuMope) [inline src/rod_mechanics/softRobots.cpp:97]: Variables should be snake case. `Curvatures --> curvatures`

File: src/rod_mechanics/softRobots.cpp [hunk 5]
```diff
@@ -57,6 +57,71 @@ void softRobots::applyInitialVelocities(int limb_idx, const vector<Vector3d> &ve
...
+    // Curvatures: the first col is limb_idx, second col is node_idx, third col is cx, fourth col is cy, fifth col is cz 
+    if (Curvatures.cols() != 4) {
+        throw runtime_error("The twist input matrix should have 3 columns!");
+    }
+
+    for (int i = 0; i < Curvatures.rows(); i++) {
+        // the first col is limb_idx, second col is node_idx, third col is x, fourth col is y, fifth col is z
```
Comment (QuantuMope) [inline src/rod_mechanics/softRobots.cpp:104]: Fifth column doesn't exist.

File: src/rod_mechanics/softRobots.cpp (hunk 4 above)
Comment (QuantuMope) [inline src/rod_mechanics/softRobots.cpp:97]: Change datatype to `Matrix<double, Dynamic, 4>`. Helps users know right away the shape of the matrix.

File: src/rod_mechanics/softRobots.cpp [hunk 6]
```diff
@@ -57,6 +57,71 @@ void softRobots::applyInitialVelocities(int limb_idx, const vector<Vector3d> &ve
...
+        
+        double theta = twists(i, 2);
+        limbs[limb_idx]->x(4*edge_idx + 3) += theta;
+    }
+}
+
+void softRobots::applyCurvatureBC(const MatrixXd &Curvatures) {
+    // Curvatures: the first col is limb_idx, second col is node_idx, third col is cx, fourth col is cy, fifth col is cz 
+    if (Curvatures.cols() != 4) {
+        throw runtime_error("The twist input matrix should have 3 columns!");
```
Comment (QuantuMope) [inline src/rod_mechanics/softRobots.cpp:100]: Wrong error message.

File: src/rod_mechanics/softRobots.cpp [hunk 7]
```diff
@@ -57,6 +57,71 @@ void softRobots::applyInitialVelocities(int limb_idx, const vector<Vector3d> &ve
...
+        if (limb_idx >= limbs.size() || node_idx >= limbs[limb_idx]->nv) {
+            throw runtime_error("Invalid limb_idx or node_idx given!");
+        }
+        
+        Vector3d pos = positions.row(i).segment(2, 3).transpose();
+        limbs[limb_idx]->x.segment(4*node_idx, 3) += pos;
+    }
+}
+
+void softRobots::applyTwistBC(const MatrixXd &twists) {
```
Comment (QuantuMope) [inline src/rod_mechanics/softRobots.cpp:78]: Can do `Matrix<double, Dynamic, 3>`.

File: src/rod_mechanics/softRobots.cpp [hunk 8]
```diff
@@ -57,6 +57,71 @@ void softRobots::applyInitialVelocities(int limb_idx, const vector<Vector3d> &ve
     }
 }
 
+void softRobots::applyPositionBC(const MatrixXd &positions) {
```
Comment (QuantuMope) [inline src/rod_mechanics/softRobots.cpp:60]: Can do `Matrix<double, Dynamic, 5>`

File: src/rod_mechanics/softRobots.cpp [hunk 9]
```diff
@@ -57,6 +57,71 @@ void softRobots::applyInitialVelocities(int limb_idx, const vector<Vector3d> &ve
...
+
+    for (int i = 0; i < positions.rows(); i++) {
+        int limb_idx = positions(i, 0);
+        int node_idx = positions(i, 1);
+        if (limb_idx >= limbs.size() || node_idx >= limbs[limb_idx]->nv) {
+            throw runtime_error("Invalid limb_idx or node_idx given!");
+        }
+        
+        Vector3d pos = positions.row(i).segment(2, 3).transpose();
+        limbs[limb_idx]->x.segment(4*node_idx, 3) += pos;
```
Comment (QuantuMope) [inline src/rod_mechanics/softRobots.cpp:74]: This is a delta position. Should rename input `positions --> delta_pos`

File: src/app.cpp [hunk 10]
```diff
@@ -81,11 +81,35 @@ class simulationManager {
...
+            if (values.cols() == 1){
+                values.transposeInPlace();
+            }
+            if (key == "position"){
+                soft_robots->applyPositionBC(values);
+            } else if (key == "twist") {
+                soft_robots->applyTwistBC(values);
+            } else if (key == "curvature") {
+                soft_robots->applyCurvatureBC(values);
+            }
```
Comment (QuantuMope) [inline src/app.cpp:99]: Use a switch statement with default handling for incorrect keys.

File: src/rod_mechanics/softRobots.h [hunk 11]
```diff
@@ -26,6 +26,9 @@ class softRobots
 
     void lockEdge(int limb_idx, int edge_idx);
     void applyInitialVelocities(int limb_idx, const vector<Vector3d>& velocities);
+    void applyPositionBC(const MatrixXd &positions);
+    void applyTwistBC(const MatrixXd &twists);
+    void applyCurvatureBC(const MatrixXd &Curvatures);
```
Comment (QuantuMope) [inline src/rod_mechanics/softRobots.h:31]: The names for these is a bit weird for me.

I think a better name is `changeTwistBar` and `changeKappaBar`

**Review by QuantuMope — CHANGES_REQUESTED** [review]

File: py_examples/helix_case/helixExample.py
Comment (DezhongT) [reply py_examples/helix_case/helixExample.py:1]: Sure. I will create a new sample file then.

File: src/rod_mechanics/softRobots.h (hunk 11 above)
Comment (DezhongT) [reply src/rod_mechanics/softRobots.h:31]: BC means boundary conditions

File: src/app.cpp (hunk 10 above)
Comment (DezhongT) [reply src/app.cpp:99]: we can not use switch for string variable

File: py_examples/helix_case/helixExample.py
Comment (DezhongT) [reply py_examples/helix_case/helixExample.py:1]: I revert it back and add a new example file.

File: py_examples/helix_case/helixExample.py (hunk 1 above)
Comment (DezhongT) [reply py_examples/helix_case/helixExample.py]: Need more discussion later

File: src/rod_mechanics/softRobots.cpp (hunk 2 above)
Comment (DezhongT) [reply src/rod_mechanics/softRobots.cpp:119]: Resolved

File: src/rod_mechanics/softRobots.cpp [hunk 12]
```diff
@@ -57,67 +57,47 @@ void softRobots::applyInitialVelocities(int limb_idx, const vector<Vector3d> &ve
...
+        limbs[limb_idx]->x(4*edge_idx + 3) += dtheta;
     }
+}
 
-    for (int i = 0; i < Curvatures.rows(); i++) {
-        // the first col is limb_idx, second col is node_idx, third col is x, fourth col is y, fifth col is z 
-        int limb_idx = Curvatures(i, 0);
-        int node_idx = Curvatures(i, 1);
+void softRobots::applyCurvatureBC(const Matrix<double, Dynamic, 4> &delta_curvatures) {
+    // Curvatures: the first col is limb_idx, second col is node_idx, third col is cx, fourth col is cy, fifth col is cz
```
Comment (QuantuMope) [inline src/rod_mechanics/softRobots.cpp:89]: This comment is still incorrect.

File: src/rod_mechanics/softRobots.cpp [hunk 13]
```diff
@@ -57,67 +57,47 @@ void softRobots::applyInitialVelocities(int limb_idx, const vector<Vector3d> &ve
...
+        int node_idx = delta_curvatures(i, 1);
         if (limb_idx >= limbs.size() || node_idx >= limbs[limb_idx]->ne || node_idx < 1) {
             throw runtime_error("Invalid limb_idx or edge_idx given!");
         }
 
-        // get material frames   
-        // Vector3d curvature = Curvatures.row(i).segment(2, 3).transpose();
-        double cx = Curvatures(i, 2);
-        double cy = Curvatures(i, 3);
+        // get material frames
```
Comment (QuantuMope) [inline src/rod_mechanics/softRobots.cpp:97]: This comment is conrrect.

File: CMakeLists.txt [hunk 14]
```diff
@@ -14,6 +14,7 @@ set(CMAKE_RUNTIME_OUTPUT_DIRECTORY "../")
 set(CMAKE_INTERPROCEDURAL_OPTIMIZATION TRUE)
 set(MAGNUM_WITH_SDL2APPLICATION ON)
 set(THREADS_PREFER_PTHREAD_FLAG ON)
+set(CMAKE_PREFIX_PATH "/home/dezhongt/.local/lib/python3.10/site-packages/pybind11/share/cmake/pybind11")
```
Comment (QuantuMope) [inline CMakeLists.txt:17]: Remove hardcoded local path.

File: py_dismech/__init__.py [hunk 15]
```diff
@@ -1 +1,3 @@
+# from .py_dismech import *
 from .py_dismech import *
+# from . import py_dismech
```
Comment (QuantuMope) [inline py_dismech/__init__.py:3]: Commented code here should be removed.

File: src/rod_mechanics/softRobots.h (hunk 11 above)
Comment (QuantuMope) [reply src/rod_mechanics/softRobots.h]: Right, but we are not changing twist or curvature to a certain boundary condition value. Instead we are changing the natural twists / curvatures, which will indirectly result in a nonlinear change in twist or curvature. This is different from positionBC, which actually changes the position to the desired value.

File: py_examples/helix_case/helixExample.py (hunk 1 above)
Comment (QuantuMope) [reply py_examples/helix_case/helixExample.py]: Sure, we can discuss offline.

**Comment by QuantuMope** [comment]
Looks good to me. Last thing is just the pre-commit change. You can just follow the commands I specified.

**Pull request merged**